File: Minerva/DotNet/DB/Info.cs

///////////////////////////////////////////////////////////////////////////////
//
//  Copyright (c) 2006, Decision Theater at Arizona State University
//  All rights reserved.
//  BSD License: http://www.opensource.org/licenses/bsd-license.html
//  Created by: Adam Kubach
//
///////////////////////////////////////////////////////////////////////////////

using Strings = System.Collections.Generic.List<string>;

namespace DT.Minerva.DB
{
  public class Info :
    DT.Minerva.DB.Connection,
    DT.Minerva.Interfaces.IDataSource
  {
    public Info()
    {
    }

    /// <summary>
    /// Get tables that contain polygon data.
    /// </summary>
    public string[] PolygonTables
    {
      get
      {
        Strings strings = new Strings();

        string[] tableNames = this.GeometryTables;

        foreach (string table in tableNames)
        {
          string geomType = this._getGeometryType(table);

          if (geomType == "POLYGON" || geomType == "MULTIPOLYGON")
          {
            strings.Add(table);
          }
        }

        return strings.ToArray();
      }
    }


    /// <summary>
    /// Get all tables that contain line data.
    /// </summary>
    public string[] LineTables
    {
      get
      {
        Strings strings = new Strings();

        string[] tableNames = this.GeometryTables;

        foreach (string table in tableNames)
        {
          string geomType = this._getGeometryType(table);

          if (geomType == "POLYGON" || geomType == "MULTIPOLYGON" || geomType == "LINESTRING" || geomType == "MULTILINESTRING")
          {
            strings.Add(table);
          }
        }

        return strings.ToArray();
      }
    }


    /// <summary>
    /// 
    /// </summary>
    public string[] PointTables
    {
      get
      {
        Strings strings = new Strings();

        string[] tableNames = this.GeometryTables;

        foreach (string table in tableNames)
        {
          string geomType = this._getGeometryType(table);

          if (geomType == "POINT" || geomType == "MULTIPOINT")
          {
            strings.Add(table);
          }
        }

        return strings.ToArray();
      }
    }


    /// <summary>
    /// Get tables that have point data and a time column.
    /// </summary>
    public string[] PointTimeTables
    {
      get
      {
        Strings strings = new Strings();

        // Get point tables.
        string[] tableNames = this.PointTables;

        foreach (string table in tableNames)
        {
          // If this table has a date column...
          if (this.hasColumnType(table, "date"))
          {
            strings.Add(table);
          }
        }

        return strings.ToArray();
      }
    }


    /// <summary>
    /// Get all tables that have a geometry column.
    /// </summary>
    public string[] GeometryTables
    {
      get
      {
        Strings strings = new Strings();

        string query = "SELECT table_name FROM information_schema.columns WHERE column_name='geom'";
        Npgsql.NpgsqlDataReader dr = this.executeQueryDataReader(query);

        while (dr.Read())
        {
          strings.Add((string)dr["table_name"]);
        }

        return strings.ToArray();

      }
    }


    /// <summary>
    /// Get all the column names of the given table.
    /// </summary>
    public string[] getColumnNames(string table)
    {
      Strings strings = new Strings();

      string query = "SELECT column_name FROM information_schema.columns WHERE table_name='" + table + "'";

      Npgsql.NpgsqlDataReader dr = this.executeQueryDataReader(query);

      while (dr.Read())
      {
        strings.Add((string)dr[0]);
      }

      return strings.ToArray();
    }


    /// <summary>
    /// Get all column names for the given table and type.
    /// </summary>
    public string[] getColumnNames(string table, string type)
    {
      Strings strings = new Strings();
      string query = "SELECT column_name FROM information_schema.columns WHERE table_name='" + table + "' AND data_type='" + type + "'";
      Npgsql.NpgsqlDataReader dr = this.executeQueryDataReader(query);

      while (dr.Read())
      {
        strings.Add((string)dr[0]);
      }

      return strings.ToArray();
    }


    /// <summary>
    /// Does the table have a column of the given type.
    /// </summary>
    public bool hasColumnType(string tableName, string type)
    {
      string[] columnNames = this.getColumnNames(tableName);

      foreach (string column in columnNames)
      {
        string query = "SELECT data_type FROM information_schema.columns WHERE table_name='" + tableName + "' AND column_name='" + column + "'";

        Npgsql.NpgsqlDataReader dr = this.executeQueryDataReader(query);

        if (dr.Read())
        {
          string columnType = (string)dr["data_type"];
          columnType = columnType.Trim();
          if (columnType == type)
            return true;
        }
      }

      return false;
    }


    /// <summary>
    /// Get the geometry type from the table name.  Assumes that the table has a column named geom.
    /// </summary>
    protected string _getGeometryType(string tableName)
    {
      string query = "SELECT GeometryType(geom) FROM " + tableName + " WHERE IsEmpty(geom)=false LIMIT 1";

      Npgsql.NpgsqlDataReader geomTypeDR = this.executeQueryDataReader(query);

      if (geomTypeDR.Read())
      {
        if (!geomTypeDR.IsDBNull(0))
        {
          return (string)geomTypeDR[0];
        }
      }

      return "";
    }


    /// <summary>
    /// 
    /// </summary>
    public string getMinValue(string tablename, string column)
    {
      string query = "SELECT MIN(" + column + ") as min FROM " + tablename;

      Npgsql.NpgsqlDataReader dr = this.executeQueryDataReader(query);

      if (dr.Read())
        return dr["min"].ToString();

      return "";
    }


    /// <summary>
    /// 
    /// </summary>
    public string getMaxValue(string tablename, string column)
    {
      string query = "SELECT MAX(" + column + ") as max FROM " + tablename;

      Npgsql.NpgsqlDataReader dr = this.executeQueryDataReader(query);

      if (dr.Read())
        return dr["max"].ToString();

      return "";
    }


    /// <summary>
    /// 
    /// </summary>
    public string[] getDistinctValues(string tablename, string column)
    {
      Strings strings = new Strings();

      string query = "SELECT DISTINCT(" + column + ") FROM " + tablename + " ORDER BY " + column;

      Npgsql.NpgsqlDataReader dr = this.executeQueryDataReader(query);

      while (dr.Read())
      {
        string value = dr[0].ToString();

        strings.Add(value.Trim());
      }

      return strings.ToArray();
    }


    /// <summary>
    /// Get all public tables.
    /// </summary>
    public string[] Tables
    {
      get
      {
        Strings strings = new Strings();

        string query = "SELECT table_name FROM information_schema.tables WHERE table_schema='public'";

        Npgsql.NpgsqlDataReader dr = this.executeQueryDataReader(query);

        while (dr.Read())
        {
          strings.Add((string)dr[0]);
        }

        return strings.ToArray();
      }
    }


    /// <summary>
    /// 
    /// </summary>
    public string getColumnType(string tableName, string column)
    {
      string query = "SELECT data_type FROM information_schema.columns WHERE table_name='" + tableName + "' AND column_name='" + column + "'";

      Npgsql.NpgsqlDataReader dr = this.executeQueryDataReader(query);

      if (dr.Read())
      {
        return dr[0] as string;
      }

      return "";
    }
  }
}

File: Minerva/Core/DB/Info.cpp

///////////////////////////////////////////////////////////////////////////////
//
//  Copyright (c) 2006, Arizona State University
//  All rights reserved.
//  BSD License: http://www.opensource.org/licenses/bsd-license.html
//  Created by: Adam Kubach
//
///////////////////////////////////////////////////////////////////////////////

#include "Minerva/Core/DB/Info.h"

using namespace Minerva::Core::DB;


///////////////////////////////////////////////////////////////////////////////
//
//  Constructor.
//
///////////////////////////////////////////////////////////////////////////////

Info::Info( Connection* connection ) : 
  BaseClass(),
  _connection ( connection )
{
}


///////////////////////////////////////////////////////////////////////////////
//
//  Destructor.
//
///////////////////////////////////////////////////////////////////////////////

Info::~Info()
{
}


///////////////////////////////////////////////////////////////////////////////
//
//  Get all public tables.
//
///////////////////////////////////////////////////////////////////////////////

Info::Strings Info::tables()
{
  std::string query ( "SELECT table_name FROM information_schema.tables WHERE table_schema='public'" );
  return this->_fillStringsFromQuery( query );
}


///////////////////////////////////////////////////////////////////////////////
//
//  Get all polygon tables.
//
///////////////////////////////////////////////////////////////////////////////

Info::Strings Info::polygonTables()
{
  Strings strings;

  Strings geometryTables ( this->geometryTables() );

  for( Strings::const_iterator iter = geometryTables.begin(); iter != geometryTables.end(); ++iter )
  {
    if( this->isPolygonTable( *iter ) )
      strings.push_back( *iter );
  }

  return strings;
}


///////////////////////////////////////////////////////////////////////////////
//
//  Get all line tables.
//
///////////////////////////////////////////////////////////////////////////////

Info::Strings Info::lineTables()
{
  Strings strings;

  Strings geometryTables ( this->geometryTables() );

  for( Strings::const_iterator iter = geometryTables.begin(); iter != geometryTables.end(); ++iter )
  {
    if( this->isLineTable( *iter ) )
      strings.push_back( *iter );
  }

  return strings;
}


///////////////////////////////////////////////////////////////////////////////
//
//  Get all point tables.
//
///////////////////////////////////////////////////////////////////////////////

Info::Strings Info::pointTables()
{
  Strings strings;

  Strings geometryTables ( this->geometryTables() );

  for( Strings::const_iterator iter = geometryTables.begin(); iter != geometryTables.end(); ++iter )
  {
    if( this->isPointTable( *iter ) )
      strings.push_back( *iter );
  }

  return strings;
}


///////////////////////////////////////////////////////////////////////////////
//
//  Get all point time tables.
//
///////////////////////////////////////////////////////////////////////////////

Info::Strings Info::pointTimeTables()
{
  Strings strings;

  Strings pointTables ( this->pointTables() );

  for( Strings::const_iterator iter = pointTables.begin(); iter != pointTables.end(); ++iter )
  {
    if( this->hasColumnType( *iter, "date" ) )
      strings.push_back( *iter );
  }

  return strings; 
}


///////////////////////////////////////////////////////////////////////////////
//
//  Get all geometry tables.
//
///////////////////////////////////////////////////////////////////////////////

Info::Strings Info::geometryTables()
{
  std::string query ( "SELECT table_name FROM information_schema.columns WHERE column_name='geom'" );
  return this->_fillStringsFromQuery( query );
}


///////////////////////////////////////////////////////////////////////////////
//
//  Get the geometry type for a table.
//
///////////////////////////////////////////////////////////////////////////////

std::string Info::_getGeometryType ( const std::string& table )
{
  std::string query ( "SELECT GeometryType(geom) FROM " + table + " WHERE IsEmpty(geom)=false LIMIT 1" );
  pqxx::result result ( _connection->executeQuery ( query ) );

  if( false == result.empty() )
    return result[0][0].as < std::string > ();

  return "";
}


///////////////////////////////////////////////////////////////////////////////
//
//  Copy results of a query into a vector.
//
///////////////////////////////////////////////////////////////////////////////

Info::Strings Info::_fillStringsFromQuery( const std::string& query )
{
  Strings strings;

  pqxx::result result ( _connection->executeQuery ( query ) );
  for( pqxx::result::const_iterator iter = result.begin(); iter != result.end(); ++iter )
  {
    strings.push_back( iter[0].as < std::string > () );
  }

  return strings;
}


///////////////////////////////////////////////////////////////////////////////
//
//  Is the given table a point table?
//
///////////////////////////////////////////////////////////////////////////////

bool Info::isPointTable ( const std::string& table ) const
{
  std::string query ( "SELECT GeometryType(geom) FROM " + table + " WHERE IsEmpty(geom)=false LIMIT 1" );
  pqxx::result result ( _connection->executeQuery ( query ) );

  if( false == result.empty() )
  {
    std::string type ( result[0][0].as < std::string > () );
    return ( ( type == "POINT" ) || ( type == "MULTIPOINT" ) );
  }

  return false;
}


///////////////////////////////////////////////////////////////////////////////
//
//  Is the given table a line table?
//
///////////////////////////////////////////////////////////////////////////////

bool Info::isLineTable ( const std::string& table ) const
{
  std::string query ( "SELECT GeometryType(geom) FROM " + table + " WHERE IsEmpty(geom)=false LIMIT 1" );
  pqxx::result result ( _connection->executeQuery ( query ) );

  if( false == result.empty() )
  {
    std::string type ( result[0][0].as < std::string > () );
    return ( ( type == "LINE" ) || ( type == "MULTILINE" ) );
  }

  return false;
}


///////////////////////////////////////////////////////////////////////////////
//
//  Is the given table a polygon table?
//
///////////////////////////////////////////////////////////////////////////////

bool Info::isPolygonTable ( const std::string& table ) const
{
  std::string query ( "SELECT GeometryType(geom) FROM " + table + " WHERE IsEmpty(geom)=false LIMIT 1" );
  pqxx::result result ( _connection->executeQuery ( query ) );

  if( false == result.empty() )
  {
    std::string type ( result[0][0].as < std::string > () );
    return ( ( type == "POLYGON" ) || ( type == "MULTIPOLYGON" ) );
  }

  return false;
}


///////////////////////////////////////////////////////////////////////////////
//
//  Does the table have a column of the given type?
//
///////////////////////////////////////////////////////////////////////////////

bool Info::hasColumnType ( const std::string& table, const std::string& type ) const
{
  std::string query ( "SELECT column_name FROM information_schema.columns WHERE table_name='" + table + "' AND data_type='" + type + "'" );
  pqxx::result result ( _connection->executeQuery( query ) );
  return !result.empty();
}
